### ISSUE TITLE
SIGHUP does not kill fish and foreground process

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#7288]
With 81d87d1c7 the pexpects/signal.py test actually *fails* when it does:

```python
# Verify that sending SIGHUP to the shell, such as will happen when the tty is
# closed by the terminal, terminates the shell and the foreground command and
# any background commands run from that shell.
send("sleep 130 &\r")
expect_prompt()
send("sleep 131 &\r")
expect_prompt()
send("sleep 132\r")
os.kill(sp.spawn.pid, signal.SIGHUP)

# Verify the spawned fish shell has exited.
sp.spawn.wait()
```

It just waits until the sleep has ended and then returns true, because spawn.wait() blocks indefinitely.

I can reproduce this by running `cat` in an interactive shell and sending SIGHUP to it. Killing the `cat` then will also end up quitting fish.

The relevant bits from strace capture:

Working fish:

```strace
read(4, 0x7ffe57547060, 4096)           = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGHUP {si_signo=SIGHUP, si_code=SI_USER, si_pid=114629, si_uid=1000} ---
getpid()                                = 114519
write(5, "\0", 1)                       = 1
rt_sigreturn({mask=[]})                 = -1 EINTR (Unterbrechung während des Betriebssystemaufrufs)
getpgrp()                               = 114519
ioctl(0, TIOCSPGRP, [114519])           = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
stat("/home/alfa/.config/fish/fish_variables", {st_mode=S_IFREG|0644, st_size=2667, ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_START or TCSETS, {B38400 opost isig -icanon -echo ...}) = 0
access("/proc/self/stat", R_OK)         = 0
openat(AT_FDCWD, "/proc/114551/stat", O_RDONLY|O_CLOEXEC) = 11
fcntl(11, F_GETFL)                      = 0x8000 (flags O_RDONLY|O_LARGEFILE)
fstat(11, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
read(11, "114551 (cat) S 114519 114551 102"..., 1024) = 314
close(11)                               = 0
getpgrp()                               = 114519
getpgrp()                               = 114519
kill(-114551, SIGHUP)                   = 0
write(1, "\33", 1)                      = 1
write(1, "[", 1)                        = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=114551, si_uid=1000, si_status=SIGHUP, si_utime=0, si_stime=0} ---
getpid()                                = 114519
```

Broken fish

```strace
--- SIGHUP {si_signo=SIGHUP, si_code=SI_USER, si_pid=116003, si_uid=1000} ---
getpid()                                = 115921
write(5, "\0", 1)                       = 1
rt_sigreturn({mask=[]})                 = -1 EINTR (Unterbrechung während des Betriebssystemaufrufs)
read(4, "\0", 4096)                     = 1
read(4, 0x7ffeb89d7110, 4096)           = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=115989, si_uid=1000, si_status=SIGINT, si_utime=0, si_stime=0} ---
```

Bisects to 04c6442dcc23567e6176203227f258d50b6018ac.

Note that disabling fish_exit handlers does not work.

**Comment by ridiculousfish** [comment]
There's a timing issue in the signals test:

    send("sleep 132\r")
    os.kill(sp.spawn.pid, signal.SIGHUP)

sometimes fish receives HUP before it has kicked off the sleep, and then the test exits quickly. Other times it receives HUP after it kicks off the sleep and only exits after the sleep process completes. Adding a Python-level sleep between the two lines resolves the race.

**Comment by faho** [comment]
Fixed by 0379f2187079c6d603109902b15a5f967cd5896f if I understand correctly?

**Comment by ridiculousfish** [comment]
Yep, forgot to close this. Thanks
